File: Day153_2025/Q153_Leetcode.py
from typing import List


class Solution:
    def candy(self, ratings: List[int]) -> int:
        n = len(ratings)
        candy = n
        i = 1

        while i < n:
            if ratings[i] == ratings[i - 1]:
                i += 1
                continue

            peak = 0
            while i < n and ratings[i] > ratings[i - 1]:
                peak += 1
                candy += peak
                i += 1

            valley = 0
            while i < n and ratings[i] < ratings[i - 1]:
                valley += 1
                candy += valley
                i += 1

            candy -= min(peak, valley)

        return candy

File: Day153_2025/Q153_Leetcode.java
class Q153_Leetcode {
    public int candy(int[] ratings) {
        int n = ratings.length;
        int candy = n, i = 1;

        while (i < n) {
            if (ratings[i] == ratings[i - 1]) {
                i++;
                continue;
            }

            int peak = 0;
            while (i < n && ratings[i] > ratings[i - 1]) {
                peak++;
                candy += peak;
                i++;
            }

            int valley = 0;
            while (i < n && ratings[i] < ratings[i - 1]) {
                valley++;
                candy += valley;
                i++;
            }

            candy -= Math.min(peak, valley);
        }

        return candy;
    }
}
